### PR TITLE
Add custom tools

### PR DESCRIPTION
**Pull request by charli117**
add gaode weather and github repository search tools

### CODE REVIEW

File: api/core/tools/provider/builtin/gaode/tools/weather.yaml [hunk 1]
```diff
@@ -0,0 +1,28 @@
+identity:
+  name: weather
```
Comment (Yeuoly) [inline api/core/tools/provider/builtin/gaode/tools/weather.yaml]: tool name should be unique to tool of other providers, it had better be "gaode_weather"

File: api/core/tools/provider/builtin/gaode/tools/weather.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+                                content['date'] = item.get('date')
+                                content['week'] = item.get('week')
+                                content['dayweather'] = item.get('dayweather')
+                                content['daytemp_float'] = item.get('daytemp_float')
+                                content['daywind'] = item.get('daywind')
+                                content['nightweather'] = item.get('nightweather')
+                                content['nighttemp_float'] = item.get('nighttemp_float')
+                                contents.append(content)
+                        s.close()
+                        return self.create_text_message(self.summary(user_id=user_id, content=json.dumps(contents)))
```
Comment (Yeuoly) [inline api/core/tools/provider/builtin/gaode/tools/weather.py]: json.dumps directly could cause an Unicode encoding problem, use ensure_ascii=False instead is a better solution
```python
try:
    return self.create_text_message(self.summary(user_id=user_id, content=json.dumps(contents, ensure_ascii=False)))
except Exception as e:
    return self.create_text_message(self.summary(user_id=user_id, content=json.dumps(contents)))
```

File: api/core/tools/provider/builtin/gaode/tools/weather.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+            City_data = city_response.json()
+            if city_response.status_code == 200 and City_data.get('info') == 'OK':
+                if len(City_data.get('districts')) > 0:
+                    CityCode = City_data['districts'][0]['adcode']
+                    weatherInfo_response = s.request(method='GET',
+                                                     url="{url}/weather/weatherInfo?city={citycode}&extensions=all&key={apikey}&output=json"
+                                                         "".format(url=api_domain, citycode=CityCode,
+                                                                   apikey=self.runtime.credentials.get('api_key')))
+                    weatherInfo_data = weatherInfo_response.json()
+                    if weatherInfo_response.status_code == 200 and weatherInfo_data.get('info') == 'OK':
```
Comment (Yeuoly) [inline api/core/tools/provider/builtin/gaode/tools/weather.py]: this `if` needs an else branch or this method will return a None

File: api/core/tools/provider/builtin/github/tools/repositories.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+                        content['owner'] = item['owner']['login']
+                        content['name'] = item['name']
+                        content['description'] = item['description'][:100] + '...' if len(item['description']) > 100 else item['description']
+                        content['url'] = item['html_url']
+                        content['star'] = item['watchers']
+                        content['forks'] = item['forks']
+                        content['updated'] = updated_at_object.strftime("%Y-%m-%d")
+                        contents.append(content)
+                    s.close()
+                    return self.create_text_message(self.summary(user_id=user_id, content=json.dumps(contents)))
```
Comment (Yeuoly) [inline api/core/tools/provider/builtin/github/tools/repositories.py]: same as above

File: api/core/tools/provider/builtin/gaode/tools/weather.yaml (hunk 1 above)
Comment (charli117) [reply api/core/tools/provider/builtin/gaode/tools/weather.yaml]: Renamed

File: api/core/tools/provider/builtin/gaode/tools/weather.py (hunk 2 above)
Comment (charli117) [reply api/core/tools/provider/builtin/gaode/tools/weather.py]: optimized

File: api/core/tools/provider/builtin/gaode/tools/weather.py (hunk 3 above)
Comment (charli117) [reply api/core/tools/provider/builtin/gaode/tools/weather.py]: optimized

File: api/core/tools/provider/builtin/github/tools/repositories.py (hunk 4 above)
Comment (charli117) [reply api/core/tools/provider/builtin/github/tools/repositories.py]: The condition here is complete

**Review by Yeuoly — DISMISSED** [review]

**Review by Yeuoly — APPROVED** [review]

**Pull request merged**